### DISCUSSION TITLE
Problems while running the Daikoku jar

### DISCUSSION THREAD

**Discussion by rupipal** [MAIF/daikoku#337] [Q&A] [answered]
Hi,

I haven't used MongoDB earlier. And presently my idea is to quickly check out daikou. When I run as
java -jar daikoku.jar It seems the connection to the DB is not set up...this is what I get in the logs...any tips?

[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] MongoDataStore - used
[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] AuditTrailPurgeJob - Start audit trail purge job
[info] AuditTrailPurgeJob - audit by cron ==> false every 1 hour
[info] r.api.Driver - No mongo-async-driver configuration found
[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] r.api.Driver - [Supervisor-1] Creating connection: Connection-1
[info] r.c.a.MongoDBSystem - [Supervisor-1/Connection-1] Starting the MongoDBSystem
[info] r.c.netty.Pack - Instantiated reactivemongo.core.netty.Pack
[info] play.api.Play - Application started (Prod) (no global state)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:8080
[warn] r.a.MongoConnection - [Supervisor-1/Connection-1] Timeout after 8200 milliseconds while probing the connection monitor: IsPrimaryAvailable#1250956357?
[warn] r.a.MongoConnection - [Supervisor-1/Connection-1] Timeout after 8200 milliseconds while probing the connection monitor: IsPrimaryAvailable#1234744190?
...

It continues like that...and , of course, the Daikoku web UI isn't there...it shows error Error - Server Error: a8dab997-9efe-3d11-8a03-2d56a53f85ed

**Comment by quentinovega** [comment, marked as answer]
Hi.
Are you sure that your mongoDB uri is `mongodb://localhost:27017/daikoku`, that's the default value but if url, port or user/password is differents, daikoku can't connect on it. 
You can override the value by passing paramater like java -jar -Dmongodb.uri=<your mongo uri> daikoku.jar

ps: I move this issue to discussion tab

**Reply by rupipal** [reply]
Hi Quentin,

Thanks for your reply. Since, possibly due to fewer users on the gitter channel, I didn't get a response, I thought of raising this as an issue. I'm using MongDb packed in with the daikoku.jar by default. I haven't done any tweaking. Moreover, i haven't even found the config files yet. If a set of uri, post and pw are already set by default, I am sorry, I am unable to understand as what to supply as options.

regards
Rupinder

**Reply by quentinovega** [reply]
There is no mongoDB packed with the jar.  You need to link an external database to daikoku,  running locally or not.
There is a big docker-compose file to run daikoku & otoroshi for a demo [here](https://github.com/MAIF/daikoku/tree/master/demo) if you want.

**Comment by rupipal** [comment]
Thanks. That's a helpful  clarification. Will check it out. I'm aware of
docker-compose file, but I am looking at a bit more detailed quick-start
that ultimately takes me to a production detail. Hence, the jar.

**Reply by quentinovega** [reply]
I understand. With the next release comes new documentation.
Daikoku can run without otoroshi or whatever, it need just a database so, you can comment all of services in docker-compose except mongoDB to start.

**Comment by rupipal** [comment]
Thanks, Quenetin. I did quite a bit of survey of API gateways and API
management tools. I find daikoku + otoroshi very attractive and one
important reason being it is fully open-source. That is, there are not
enterprise-edition features. I've faced an issue with helm chart for
otoroshi too. I will get to it later at the otoroshi repository. YES,
documentation needs updates. Speaking of otoroshi, it needs updates for
K8s, e.g. minikube. Fo example, I'm installing on minikube; so the issue
with the helm chart could be due to minikube.

If I continue to enjoy some support here and otoroshi, I'd like to stick to
it. BTW, gitter channels have only some users; that could be the reason of
very slow response there :)

**Comment by rupipal** [comment]
Hi Quentin,

Finally, I decided to use postgresql with daikoku. I have overridden a few
daikoku storage  defaults because they do not exist for my postgresql
installation. E.g., there is no "default" database. It is called
"postgres". And since postgres user/roles on my installation don't seem to
use passwords, I created another role called postgres2 with password
postgres2 and given it as parameters. Unluckily, I still land at this :) Is
it because I didn't have the otoroshi instance up; also, it is currently
created with a fileDB only.


java -jar -Ddaikoku.storage=postgres -Ddaikoku.postgres.database=postgres
-Ddaikoku.postgres.username=postgres2 -Ddaikoku.postgres.password=postgres2
daikoku.jar

[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] PostgresDataStore - used : postgres
[info] AuditTrailPurgeJob - Start audit trail purge job
[info] AuditTrailPurgeJob - audit by cron ==> false every 1 hour
[info] play.api.Play - Application started (Prod) (no global state)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:8080
[error] otoroshi-reactive-pg-kv - Failed to apply query: "SELECT EXISTS
(SELECT FROM information_schema.tables WHERE table_schema = $1 AND
table_name = 'tenants')" with params: "public"
[error] otoroshi-reactive-pg-kv - java.util.concurrent.ExecutionException:
Boxed Exception
[error] otoroshi-reactive-pg-kv - Failed to apply query: "SELECT EXISTS
(SELECT FROM information_schema.tables WHERE table_schema = $1 AND
table_name = 'tenants')" with params: "public"
[error] otoroshi-reactive-pg-kv - java.util.concurrent.ExecutionException:
Boxed Exception
...

Rupinder

**Reply by Zwiterrion** [reply]
Hi,

I just tested it, with the v1.5-alpha.8 version and it's working. 

I run with :

```sh
java -jar \
-DDAIKOKU_POSTGRES_HOST=localhost \
-DDAIKOKU_POSTGRES_DATABASE=postgres \
-DDAIKOKU_STORAGE=postgres \
-DDAIKOKU_POSTGRES_PASSWORD=password \
-DDAIKOKU_POSTGRES_USERNAME=postgres \
-DDAIKOKU_POSTGRES_SCHEMA=public \
-DDAIKOKU_POSTGRES_PORT=5432 \
daikoku.jar
```
and with the version of java 

```sh
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)
```

My postgres is running in a Docker container. Nothing to different from the default configuration. 

```sh
docker run -p 5432:5432 -e POSTGRES_PASSWORD=password -d --name postgres postgres
```

Let us know if after that, it continue to not working.

**Comment by rupipal** [comment]
Hi Etienne,

Thanks for your reply.
"postgres" is a superuser by default with no password on my postgresql
installation on Linux Debian 11; the installation was done with apt-get.
So, I'd set the new role postgres2 with a password. I have now re-created
this role with due care. I will write those steps here too so that we tally
on this (I've mostly used Apache Derby that comes bundled with GlassFish).

CREATE ROLE postgres2 INHERIT LOGIN CREATEDB CREATEROLE PASSWORD 'secret';

And then I've given some more privileges to it by granting some predefined
roles in postgresql.

GRANT pg_read_all_data TO postgres2;

GRANT pg_write_all_data TO postgres2;

(does it need any more privileges ?)

And the role is duly created,
\du

Role name |                         Attributes                         |
           Member of
-----------+------------------------------------------------------------+--------------------------------------
 postgres2 | Create role, Create DB                                     |
{pg_read_all_data,pg_write_all_data}


And, then, I I'm still stuck with the same errors,


java -jar -Ddaikoku.storage=postgres -Ddaikoku.postgres.database=postgres
-Ddaikoku.postgres.username=postgres2 -Ddaikoku.postgres.password=secret
-Ddaikoku.postgres.port=5432 -Ddaikoku.postgres.host=localhost
-Ddaikoku.postgres.schema=public daikoku.jar
[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] a.e.s.Slf4jLogger - Slf4jLogger started
[info] PostgresDataStore - used : postgres
[info] AuditTrailPurgeJob - Start audit trail purge job
[info] AuditTrailPurgeJob - audit by cron ==> false every 1 hour
[info] play.api.Play - Application started (Prod) (no global state)
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:8080
[error] otoroshi-reactive-pg-kv - Failed to apply query: "SELECT EXISTS
(SELECT FROM information_schema.tables WHERE table_schema = $1 AND
table_name = 'tenants')" with params: "public"
[error] otoroshi-reactive-pg-kv - java.util.concurrent.ExecutionException:
Boxed Exception
[error] otoroshi-reactive-pg-kv - Failed to apply query: "SELECT EXISTS
(SELECT FROM information_schema.tables WHERE table_schema = $1 AND
table_name = 'tenants')" with params: "public"


Passing -Ddaikoku.postgres.port=5432,  -Ddaikoku.postgres.host=localhost,
-Ddaikoku.postgres.schema=public  was redundant since they are already
daikoku defaults; but I've done it.


regards
Rupinder

**Comment by rupipal** [comment]
Hi @Zwiterrion @quentinovega  any more tips :)

**Reply by quentinovega** [reply]
Hi. Did you try with user postgres just to test ? I don't see anything abnornal but...why not because with default values, it work fine for us.
by the way, can you give us your java version and postgres version please ?

**Reply by rupipal** [reply]
Hi Quentin,
Thanks for your reply. Now with freshly downloaded jar, I've tested with default postgresql role postgres and pw postgres. I'm getting the same errors as before.  My java is openjdk 11and postgresql is 14.

regards
Rupinder

**Reply by rupipal** [reply]
Also I took care to set a password for the default user/role 'postgres', which by default doesn't have one.

**Reply by Zwiterrion** [reply]
Hi @rupipal,

I just tested with a postgres in version `14-alpine` and I successfully reproduced the bug. 

We will deployed soon a `v1.5-alpha.9` to fix it. 

For the moment, you can try with a version less than 14.

**Comment by rupipal** [comment]
Thanks Etienne :) will try with <14.

**Comment by rupipal** [comment]
Hi Etienne,

I've checked with postgresql 13. It works :) Also, on
https://maif.github.io/daikoku/devmanual/firstrun/configfile.html  , in the
section postgres configuration, daikoku.postgres.password won't have
anything like default password 'postgres.' The user may set whatever they
like. accordingly, this may read something like 'your secret'


regards
Rupinder

**Reply by Zwiterrion** [reply]
Hi Rupipal,

Glad to read this. 
I'll check the default configuration, as it is injected when Daikoku starts up and merged with the configuration given by the user.

The default postgres configuration is in the `base.conf` file.
```sh
postgres {
      host = "localhost"
      schema = "public"
      database = "default"
      username = "postgres"
      password = "postgres"
      port = 5432
      poolSize = 30
      ssl {
        enabled = false
        mode = "verify-ca"
      }
}
```

**Comment by rupipal** [comment]
Hi Etienne,

I need another tip. My Otoroshi instance is running on port 8080 from a
fileDB. It has a couple of default APIs plus another quickstart service
that I set up. The daikoku instance is running on 9090 from postgresql.
Following this,
https://maif.github.io/daikoku/devmanual/tenantusage/1-otoroshi.html  , I
'm trying to see the otoroshi instance. It doesn't show anything. I just
see the wait/checking prompt. What am I missing?


regards
Rupinder

**Reply by quentinovega** [reply]
Hi,
the manual could be rough sometimes ;) 
 You have to fill the form with: 

1. the otoroshi-api url (ex: http://otoroshi-api.oto.tools:8080)
2. the host (ex: otoroshi-api.oto.tools)
3. the client id of an admin apikey of otoroshi => you can find it on the otoroshi UI
4. the client secret of an admin apikey of otoroshi => you can find it on otoroshi UI

regards
Quentin

**Comment by rupipal** [comment]
Hi Quentin,

Thanks :)  I've created an otoroshi instance setting with admin api things.
Unluckily, I still can't get that initialized in the default tenant. Again,
I keep seeing the wait/checking prompt. Tried logging-out/logging-in and
looked around the manual.

regards
Rupinder

**Comment by rupipal** [comment]
Hi Quentin,
BTW, though it doesn't show otorishi instances in Initialization, I think
the instances are initialized anyway. I saw/used it while creating an API.
It is visible when generating Otoroshi APIkey. And I checked in the
Ortoroshi instance; it had indeed generated an API key.

regards
Rupinder

**Reply by quentinovega** [reply]
Hi Rupinder,
Ok I see. It's a bug due to our recent migration to graphql..
it's fixed and it will be accessible in the next release.
Thanks

**Comment by rupipal** [comment]
Thanks Quentin. Would be back if I have some important query :)
